ML's architecture change - Unit Testing and Proxy changes

rabarona opened the pull request:
-Fixed distinctDocumnt RDD[String]. It is being collected in multiple places, added cache() to have same list in same order every time.
-Replacing LDAWrapper output from Map[String, Array[String]] to tuple (Array[String], Array[String])
-Removing unused imports from DNSWordCreationTest
-Changed ml_ops.sh parameters to be more descriptive
-Added LDAWrapperTest
-Added Flow and DNS WordCreationTest
-Merged Proxy and modified ProxyLDA to receive tuple (Array[String], Array[String]) from LDAWrapper instead of Map(String, Array[String])

NathanSegerlind committed:
mpi_cmd

MPI execution now controlled by MPI_PREP_CMD and MPI_CMD.

MPI_PREP_CMD: command necessary to prepare for running MPI (loading of environment variables or somesuch)
MPI_CMD : command to execute MPI

NathanSegerlind committed:
ml_CLI_based_parameters

- ML spark code now takes parameters as CLI passed from ml_ops.sh

- NOW and HENCEFORTH no ML pipeline (save ml_ops.sh) may access an environment variable - all parameters are explicitly passed

- ml_ops.sh : conf file gets sourced only once, nothing gets exported

- ml_ops.sh : restructuring so that pipeline parameters get defined at top and used throughout

- sleeps replaced by waits

- installation script now excludes big files not needed to run solution (eg. no more slamming around the src tree and all the stuff in target)

NathanSegerlind committed:
Merge pull request #30 from NathanSegerlind/mpi_cmd

mpi_cmd

NathanSegerlind committed:
Merge branch 'ml_structure_s13' into ml_CLI_based_parameters

NathanSegerlind committed:
Merge pull request #31 from NathanSegerlind/ml_CLI_based_parameters

ml_CLI_based_parameters

NathanSegerlind committed:
proc_count_to_conf

ml_ops.sh now picks up PROCESS_COUNT (the number of processes running the task in MPI) from duxbay.conf

User shold no longer have to edit the ml_ops.sh file when installing ML pipeline or reconfiguring MPI.

NathanSegerlind committed:
ml_structure_s13

updated README.md to reflect changes to use of duxbay.conf to configure MPI

Commit:
Changing Flow ML source, will read parquet avro now instead of csv.

NathanSegerlind committed:
Merge pull request #32 from NathanSegerlind/proc_count_to_conf

proc_count_to_conf

NathanSegerlind committed:
Merge pull request #33 from NathanSegerlind/ml_structure_s13

ml_structure_s13

Commit:
Merge remote branch 'master-repo/ml_structure_s13' into 1.1-flow_avro_as_source

Commit:
Added  folder to each node

Commit:
Changing Flow ML source, will read parquet avro now instead of csv. (#34)

NathanSegerlind committed:
Merge pull request #35 from rabarona/1.1-ml_split_result_folders

Modified LPATH to split results into each data source - ml_ops.sh

Commit:
Fixed indexes for SOURCEWORD and DESWORD

Commit:
fixed SOURCEWORD and DESTWORD indexes

Commit:
Fix LDA_OUTPUT_DIR and ml_ops.sh logic to go back to LUSER/ml folder (#38)

* Making consistent switch to /ml folder, avoid cd..

* Fixed LDA_OUPUT_DIR to include DSORUCE

Commit:
Merge remote branch 'master-repo/ml_structure_s13' into 1.1-flow_avro_as_source

Commit:
Added missing columns required by Flow OA.

Commit:
Added LDAWrapper

Commit:
Merge remote-tracking branch 'master-repo/ml_structure_s13' into 1.1-ml_lda_wrapper

Commit:
Merge branch '1.1-flow_avro_as_source' into 1.1-ml_lda_wrapper

Commit:
Move ml results to hdfs (#39)

Commit:
Initial ldawrapper changes

Commit:
Merge remote-tracking branch 'master-repo/ml_structure_s13' into 1.1-ml_lda_wrapper

Commit:
LDA wrapper code

Commit:
LDA wrapper completed. Need to validate data

Commit:
Code refactor

Commit:
Added UnitTesting for FlowWordCreation

Commit:
Code refactoring/cleanup - LADWrapper

Commit:
Added DNSWordCreationTests. Fixed small unreachable code in DNSWordCreation. Using more explanatory parameter names in ml_ops.sh

Commit:
Removed unused files -python scripts and Dispatcher.scala. Also modified LDA.scala with SuspiciousConnects.scala for a more descriptive name. Made the required change in ml_ops.sh to call --class org.opennetworkinsight.SuspiciousConnects

Commit:
updated install.sh to stop including *.py files in installation to each node.

Commit:
-Replacing LDAWrapper output from Map[String, RDD[String]] to tuple (RDD[String], RDD[String])
-Removing unused imports from DNSWordCreationTest
-Changed ml_ops.sh parameters to be more descriptive
-Added LDAWrapperTests

NathanSegerlind commented on lda_post.py:
```diff
@@ -1,123 +0,0 @@
-import numpy as np
```
good to finally clean this out!

NathanSegerlind commented on lda_pre.py:
```diff
@@ -1,95 +0,0 @@
-#!/user/bin/python
```
good to finally clean this out

NathanSegerlind commented on ml_ops.sh:
```diff
@@ -74,13 +74,13 @@ hdfs dfs -rm -R -f ${HDFS_SCORED_CONNECTS}
 # Pre MPI command can be configured in /etc/duxbay.conf
-# In this script, after the line after --mpicmd ${MPI_CMD} add:
-# --mpiprep ${MPI_PREP_CMD}
+# In this script, after the line after -c ${MPI_CMD} add:
```
I like using the verbose names better in the comment...

actually,  what does it hurt to add

--mpiprep ${MPI_PREP_CMD}

since a properly written conf file will have an empty command there ?  or does the spark options handler not allow empty strings arguments (i think it does, I mean, it _should_)

rabarona replied:
Sorry I overlooked the verbose name for that parameter. 
I couldn't get .jar to work with empty parameter but maybe I didn't try hard.

Will work in a solution.

rabarona replied:
Didn't find a quick solution for empty parameters. I will keep looking, I hope this commit make it to dev anyway.

NathanSegerlind commented on src/main/scala/org/opennetworkinsight/DNSWordCreation.scala:
```diff
@@ -20,13 +33,7 @@ import scala.math._
       var numparts = spliturl.length
       var domain = "None"
       var subdomain = "None"
-      //var country ="None"
```
what's with these changes? l24-l29 and l53-l55

rabarona replied:
While writing unit testing I found there is no uses for host variable. Removed carefully.

NathanSegerlind commented on ml_ops.sh:
```diff
@@ -74,13 +74,13 @@ hdfs dfs -rm -R -f ${HDFS_SCORED_CONNECTS}
 
 
-time spark-submit --class "org.opennetworkinsight.LDA" --master yarn-client --executor-memory  ${SPK_EXEC_MEM} \
+time spark-submit --class "org.opennetworkinsight.SuspiciousConnects" --master yarn-client --executor-memory  ${SPK_EXEC_MEM} \
```
I like this name... we should always try to use names that match WHAT the function does not HOW it does it... in this particular case, LDA is just one step in a larger "suspicious connects" analysis

NathanSegerlind commented on src/main/scala/org/opennetworkinsight/DNSLDA.scala:
```diff
@@ -15,12 +15,12 @@ object DNSLDA {
       sqlContext, logger)
 
-    val ldaResult = LDAWrapper.runLDA(docWordCount, config.modelFile, config.topicDocumentFile, config.topicWordFile,
+    val (documentResults, wordResults) = LDAWrapper.runLDA(docWordCount, config.modelFile, config.topicDocumentFile, config.topicWordFile,
```
the hard core Scala stylist would argue for case class over a tuple here...

rabarona replied:
Better than the first approach though. Will review and change accordingly.

NathanSegerlind commented on src/main/scala/org/opennetworkinsight/SuspiciousConnects.scala:
```diff
@@ -22,19 +23,21 @@ object LDA {
+          }
+          case "dns" => {
+            DNSLDA run(config, sparkContext, sqlContext, logger)
+          }
```
this is missing the proxy analysis !!

NathanSegerlind commented on src/main/scala/org/opennetworkinsight/SuspiciousConnects.scala:
```diff
@@ -22,19 +23,21 @@ object LDA {
-          case "dns" =>  DNSLDA.run(config, sparkContext, sqlContext, logger)
-          case "proxy" => ProxyLDA.run(config, sparkContext, sqlContext, logger)
-          case _ => println("ERROR:  unsupported (or misspelled) anlaysis: " + dataSource)
+          case "flow" => {
```
the brackets strike me as unnecessary for a single function call

NathanSegerlind commented on src/test/scala/org/opennetworkinsight/DNSWordCreationTest.scala:
```diff
@@ -0,0 +1,148 @@
+
+class DNSWordCreationTest extends TestingSparkContextFlatSpec with Matchers{
+
+  val countryCodesSet = Set("ac", "ad", "ae", "af", "ag", "ai", "al", "am", "an", "ao", "aq", "ar", "as", "at", "au",
```
is this duplicate code with what appears in DNSWordCreation?

we could move it to a separate file and grab it from there

NathanSegerlind commented on src/test/scala/org/opennetworkinsight/DNSWordCreationTest.scala:
```diff
@@ -0,0 +1,148 @@
+    "tl", "tm", "tn", "to", "tp", "tr", "tt", "tv", "tw", "tz", "ua", "ug", "uk", "us", "uy", "uz", "va", "vc", "ve",
+    "vg", "vi", "vn", "vu", "wf", "ws", "ye", "yt", "za", "zm", "zw")
+
+  "extractSubdomain" should "return domain=None, subdomina= None, subdomain lenght= 0 and number of parts = 6" in {
```
some typos in the test message:  subdomina=>subdomain, lenght=>length

NathanSegerlind commented on src/test/scala/org/opennetworkinsight/DNSWordCreationTest.scala:
```diff
@@ -0,0 +1,148 @@
+    result(3) shouldBe "3"
+  }
+
+  "extractSubdomain" should "return domain=None, subdomina= None, subdomain lenght= 0 and number of parts = 2" in {
```
subdomain, length

NathanSegerlind commented on src/test/scala/org/opennetworkinsight/LDAWrapperTest.scala:
```diff
@@ -0,0 +1,153 @@
+
+
+  "normalizeWord" should "calculate exponential of each value in the input string, then sum up all the exponential and " +
+    "then divided  each exponential by the total sum" in {
```
divide

NathanSegerlind commented on src/test/scala/org/opennetworkinsight/LDAWrapperTest.scala:
```diff
@@ -0,0 +1,153 @@
+    "total count for each word" in {
+
+    val documentWordData = sparkContext.parallelize(Array(Array("192.168.1.1", "333333_7.0_0.0_1.0", "8"),
+      Array("10.10.98.123", "1111111_6.0_3.0_5.0", "4"),
```
this isn't the originating location in the codebase... but low priority item (but a good one for readability) is that the binning numbers should be INTs not DOUBLEs for the word creation... the extra  ".0"s throughout the word strings hinder readability

Commit:
Merged Proxy initial code. Updated ProxyLDA to get tuple(Array[String], Array[String]  from LDAWrapper and not Map[String, Array[String]]

Commit:
Fix distinctDocument RDD[String].
distinctDocument is being collected in multiple places, added cache() to get values in the same order every time.

Commit:
Changes for @NathanSegerlind comments.

NathanSegerlind commented on src/main/scala/org/opennetworkinsight/SuspiciousConnects.scala:
```diff
@@ -23,21 +23,19 @@ object SuspiciousConnects {
-          }
+          case "flow" => FlowLDA.run(config, sparkContext, sqlContext, logger)
+          case "dns" =>  DNSLDA.run(config, sparkContext, sqlContext, logger)
+          case "proxy" => ProxyLDA.run(config, sparkContext, sqlContext, logger)
```
:+1:

NathanSegerlind replied:
:oniguy:

NathanSegerlind commented:
lookin good to me!!!

natedogs911 commented:
LGTM

EverLoSa commented:
LGTM
